### DISCUSSION TITLE
Is it possible to get dynamically generated phone numbers work in Gatsby ?

### DISCUSSION THREAD

**Discussion by VernerKeel** [gatsbyjs/gatsby#29209] [Help]
Hi, 

I'm trying to implement [Response Tap's](https://www.responsetap.com/) phone call tracking system on my Gatsby website. To track calls the system dynamically generates a unique phone number for each visitor. 
I got the system working but it only generates a dynamic number if I refresh the page.  Meaning the dynamically generated number appears after initial load but when navigating to another page then the default static number is displayed. When the page is refreshed then the dynamic number appears again. **Does anybody know how could I get it work on Gatsby or is it impossible?** 

**_Code in component_**
```
<span className='rTapNumber432151'>0161 000 0000</span>`
```
**_Script before the closing body tag in html.js_**
```
      <script
          type='text/javascript'
          dangerouslySetInnerHTML={{
            __html: `
   var adiInit = "76225", adiRVO = true;
   var adiFunc = null;
   (function() {
      var adiSrc = document.createElement("script"); adiSrc.type = "text/javascript";
      adiSrc.async = true;
      adiSrc.src = ("https:" == document.location.protocol ? "https://static-ssl" : "http://static-cdn")
      	+ ".responsetap.com/static/scripts/rTapTrack.min.js";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(adiSrc, s);
   })();`,
          }}
        />
```

**Comment by herecydev** [comment]
You need to hoist the component up otherwise state will be lost between pages. Take a look at https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#wrapPageElement

**Reply by VernerKeel** [reply]
Thank you for pointing me to the right direction. It works but I cannot figure out **how to hoist an individual button up inside the navbar?**.  
I tried to follow instructions provided [here](https://www.freecodecamp.org/news/keeping-state-between-pages-with-local-state-in-gatsby-js/) and got near perfect but not ideal outcome as it suits for separate components that have no other dynamic elements on them. 

**When I applied the below then everything worked apart from the fact than menu dropdown hover effect froze on navbar:**

```
// navbarWrapper.js

import React from 'react';

import Navbar from './navbar';

const navbarWrapper = ({ children }) => (
  <>
    <Navbar />
    {children}
  </>
);

export default navbarWrapper;
```

```
//gatsby-browser.js
import React from 'react';
import NavbarWrapper from './src/components/navbarWrapper';

export const wrapPageElement = ({ element, props }) => (
  <NavbarWrapper {...props}>{element}</NavbarWrapper>
);
```

**Reply by herecydev** [reply]
This looks correct to me, what exactly is not working?

**Reply by VernerKeel** [reply]
The call button is on a navbar that has a megamenu dropdown. When the navbar is hoisted up then when clicking on pages on the megamenu dropdown, then the dropdown doesn't disappear as it should. Here's the issue on a test page: https://test.webcontext.uk/

**Reply by herecydev** [reply]
Can you post your repo? Happy to help but it's hard to diagnose without seeing the code

**Reply by VernerKeel** [reply]
Yes, of course - here you go: https://github.com/vernerkeel/gatsby-testing-ground
Thank you for looking into it. I truly appreciate it.

**Reply by herecydev** [reply]
https://github.com/vernerkeel/gatsby-testing-ground/blob/live/src/components/megamenu/callEmailBtns.js#L10

Onclick needs to be a function, start with `onClick={() => { console.log("clicked") }` and make sure that works, then you can build on it

**Reply by VernerKeel** [reply]
Thanks for pointing that out. I've fixed it but that doesn't resolve the main issue. If all navbar is hoisted up then megamenu dropdown stays visible after clicking on it's links. **Is there a way how to hoist up only a particular button on a navbar?**

**Comment by ashleynexvelsolutions** [comment]
@VernerKeel Did you ever get this sorted out? I'm having the same problem.
